### PR TITLE
Fix memory leak and minor refactoring

### PR DESCRIPTION
**Pull request by pjueon**
- there was no `free` for the `strdup` in the code 
  - [description on strdup from cppreference: ](https://en.cppreference.com/w/c/experimental/dynamic/strdup) `The returned pointer must be passed to free to avoid a memory leak.`
  - replaced `strdup` with `std::string` (fix memory leak)
  
- minor refactoring 
  - changed time_source type to `std::unque_ptr` (remove potential memory leak risk)

### CODE REVIEW

**Comment by pjueon** [comment]
@hzeller

File: src/txtempus.cc [hunk 1]
```diff
@@ -157,7 +158,7 @@ int main(int argc, char *argv[]) {
       ttl = atoi(optarg);
       break;
     case 's':
-      service = strdup(optarg);
+      service = optarg;
```
Comment (hzeller) [inline src/txtempus.cc:161]: I think we can even avoid assigning the service string here and just directly call CreateTimeSourceFromName() here.

So in line 140, we'd have
```
   std::unique_ptr<TimeSignalSource> time_source;
  //...
  case 's':
     time_source = CreateTimeSourceFromName(optarg);
```
That way, we don't even have to worry about the service string

(Also, I's probably just leave CreateTimeSourceFromName() as is, returning a raw pointer, and assign it here as `time_source.reset(CreateTimeSourceFromName(optarg))`)

**Review by hzeller — COMMENTED** [review]
Looks good, but I'd simplify it a bit

   * Don't even use the `std::string` service, just directly create and assign time source right then
   * Avoid the repetetive `std::unique_ptr<TimeSignalSource>` in the factory, but just have that factory return the raw pointer as before. Just use `time_source.reset(...)` at the place of usage.

**Comment by pjueon** [comment]
Thanks for the feedback.

Removing the variable `service` looks good to me. 

The factory function that returns the raw pointer feels still wrong to me though. You might reuse the the factory somewhere else in the future. 

You can use `std::make_unique` to avoid repetetive `std::unique_ptr<TimeSignalSource>` in the factory:

```cpp
std::unique_ptr<TimeSignalSource> CreateTimeSourceFromName(const char *n) {
  if (strcasecmp(n, "DCF77") == 0)
    return std::make_unique<DCF77TimeSignalSource>();
  if (strcasecmp(n, "WWVB") == 0)
    return std::make_unique<WWVBTimeSignalSource>();
  if (strcasecmp(n, "JJY40") == 0)
    return std::make_unique<JJY40TimeSignalSource>();
  if (strcasecmp(n, "JJY60") == 0)
    return std::make_unique<JJY60TimeSignalSource>();
  if (strcasecmp(n, "MSF") == 0)
    return std::make_unique<MSFTimeSignalSource>();
  return nullptr;
}
```

To use `std::make_unique`, C++14 or higher is required.  But you can implement `make_unique` function if you don't want to change the C++ standard requirements:

```cpp
template<class T>
std::unique_ptr<T> make_unque(){
  return std::unique_ptr<T>{new T()};
}
```

**Comment by pjueon** [comment]
Removed `std::string service`. Please let me know your opinion about `make_unique`.
@hzeller

**Comment by hzeller** [comment]
I think switching to c++14 and using `make_unique` sounds good to me.

**Comment by hzeller** [comment]
Thanks, merged!

**Pull request merged**